flight/mcu/lib/FifoBuffer/FifoBuffer.h

```diff
@@ -1,67 +1,163 @@
+            sum += access_value(*items[i]);
+        }
+        return sum / (double) (end - start);
+    }
+
+    double getSecondDerivativeAverage(double (*access_value)(T&), systime_t (*access_time)(T&), size_t start, size_t end) {
+        T* items[end - start];
+        readSlice(items, start, end);
 
-    // returns false on failure
-    bool pop(void* out);
+        double derivatives[end - start - 1];
+        for (size_t i = start; i < end - 1; i++) {
+            double first = access_value(*items[i]);
+            double second = access_value(*items[i + 1]);
+            systime_t delta_t = access_time(*items[i + 1]) - access_time(*items[i]);
+            derivatives[i] = (second - first) / (delta_t == 0 ? 0.02 : delta_t);
+        }
 
-   private:
-    uint16_t capacity_, cur_length_, head_idx_, tail_idx_, data_size_;
-    void* arr_;
+        double second_derivatives[end - start - 2];
+        for (size_t i = start; i < end - 2; i++) {
+            double first = derivatives[i];
+            double second = derivatives[i + 1];
+            systime_t delta_t = access_time(*items[i + 1]) - access_time(*items[i]);
+            derivatives[i] = (second - first) / (delta_t == 0 ? 0.02 : delta_t);
+        }
 
-#ifdef COMPILE_LOCAL
-    std::mutex lock_;
-#endif
+        double sum = 0.0;
+        for (size_t i = start; i < end - 2; i++) {
+            sum += second_derivatives[i];
+        }
+        return sum / (double) (end - start - 2);
+    }
 
-#ifdef COMPILE_TARGET
-    mutex_t lock_;
-#endif
+private:
+    /**
+     * @brief index of the next slot to write to
+     */
+    size_t tail_idx = 0;
+    /**
+     * @brief total number of items that have been pushed
+     */
+    size_t count = 0;
+    T arr[max_size]{};
 };
 
-template <typename T, size_t max_size>
-class FifoBuffer {
-    // static_assert(std::is_trivially_copyable<T>::value, "Only trivially
-    // copyable types are allowed");
-   public:
-    FifoBuffer() : buffer(arr, max_size, sizeof(T)) {}
+template<typename T, size_t max_size>
+class FifoView {
+public:
+    friend class FifoBuffer<T, max_size>;
 
-    bool push(T element) { return buffer.push(&element); }
+    explicit FifoView(FifoBuffer<T, max_size>& observed_) : observed(observed_) {}
 
-    bool pop(T* out) { return buffer.pop(out); }
+    bool next(T& out) {
+        chMtxLock(&observed.lock);
+        if (observed.count == read_count) {
+            chMtxUnlock(&observed.lock);
+            return false;
+        }
 
-   private:
-    GenericFifoBuffer buffer;
-    T arr[max_size]{};
+        while (read_count + max_size < observed.count) {
+            read_count++;
+            idx++;
+            if (idx == max_size) {
+                idx = 0;
+            }
+        }
+
+        out = observed.arr[idx++];
+        if (idx == max_size) {
+            idx = 0;
+        }
+        read_count++;
+
+        chMtxUnlock(&observed.lock);
+        return true;
+    }
```

Comment:
what is the utility of this vs the read slice function of fifo buffer. It is also not thread safe, other threads can modify the underlying buffer, making the elements that the iterator reads out of order

Reply:
A FifoView is designed such that if you call `next` twice, you shouldn't get the same value twice, which would be useful in the SDLogger so that if the SDLogger thread goes too fast, it doesn't start logging the same data points again. It's also used in telemetry.

That's the theory anyways, who knows if it works as intended.

Reply:
makes sense, I think the code is sound